Framework/API/src/DetectorInfo.cpp

```diff
@@ -128,66 +160,102 @@ void DetectorInfo::clearMaskFlags() {
-    if (m_sample)
-      m_samplePos = m_sample->getPos();
+    const auto &detIndices = getAssemblyDetectorIndices(comp);
+    if (detIndices.size() == 0 || detIndices.size() == size()) {
+      // Update only if comp is not a bank (not detectors) or the full
+      // instrument (all detectors). The should make this thread-safe for
+      // practical purposes such as moving each bank in a separate thread.
+      if (m_source)
+        m_sourcePos = m_source->getPos();
+      if (m_sample)
+        m_samplePos = m_sample->getPos();
+      if (m_source && m_sample)
+        m_L1 = m_sourcePos.distance(m_samplePos);
+    }
     // Detector positions are currently not cached, the cached pointers to
     // detectors stay valid. Once we store positions in DetectorInfo we need to
     // update detector positions here.
+    const auto delta = toVector3d(pos - oldPos);
+    for (const auto index : detIndices) {
+      m_detectorInfo.setPosition(index, m_detectorInfo.position(index) + delta);
+    }
   }
 }
 
-/** Set the absolute rotation of the component `comp`. Not thread safe.
+/** Set the absolute rotation of the component `comp`.
  *
  * This may or may not be a detector. Even if it is not a detector it will
- * typically still influence detector positions rotations. */
+ * typically still influence detector positions rotations. Note that this method
+ * will be removed once ComponentInfo::setRotation is available. */
 void DetectorInfo::setRotation(const Geometry::IComponent &comp,
                                const Kernel::Quat &rot) {
-  using namespace Geometry::ComponentHelper;
-  TransformType rotationType = Absolute;
-  rotateComponent(comp, *m_pmap, rot, rotationType);
-
-  if (!dynamic_cast<const Geometry::Detector *>(&comp)) {
+  if (const auto *det = dynamic_cast<const Geometry::Detector *>(&comp)) {
+    const auto index = indexOf(det->getID());
+    setRotation(index, rot);
+  } else {
+    // This will go badly wrong if the parameter map in the component is not
+    // identical to ours, but there does not seem to be a way to check?
+    const auto pos = toVector3d(comp.getPos());
+    auto invOldRot = comp.getRotation();
+    invOldRot.inverse();
+    const auto delta = toQuaterniond(rot * invOldRot).normalized();
+    using namespace Geometry::ComponentHelper;
+    TransformType rotationType = Absolute;
+    rotateComponent(comp, *m_pmap, rot, rotationType);
     // If comp is a detector cached positions and rotations stay valid. In all
-    // other cases/ (higher level in instrument tree, or other leaf component
+    // other cases (higher level in instrument tree, or other leaf component
     // such as sample or source) we flush all cached positions and rotations.
-    if (m_source)
-      m_sourcePos = m_source->getPos();
-    if (m_sample)
-      m_samplePos = m_sample->getPos();
+    const auto &detIndices = getAssemblyDetectorIndices(comp);
+    if (detIndices.size() == 0 || detIndices.size() == size()) {
```

Comment:
The instrument need not be the immediate parent. This assumption would be wrong if there is an assembly that contains some(!) of the detectors and the source or sample. We do not have any guarantees, but this seem very unlikely. Without this check, we have issues with tread safety -- `SCDCalibratePanels` is moving each bank in a separate thread.

Given that we intend to remove this position chaching as soon as we have `ComponentInfo`, I hope it is ok to take this risk for now?